paddle/fluid/framework/variable.h

```diff
@@ -101,8 +110,48 @@ class Variable {
   };
 
   // pointers to a PlaceholderImpl object indeed.
-  std::unique_ptr<Placeholder> holder_;
+  std::shared_ptr<Placeholder> holder_;
 };
 
+inline framework::TensorInplaceVersion& Variable::InplaceVersionCounter() {
+  if (IsType<framework::Tensor>()) {
+    return GetMutable<framework::Tensor>()->InplaceVersionCounter();
+  } else if (IsType<framework::LoDTensor>()) {
+    return GetMutable<framework::LoDTensor>()->InplaceVersionCounter();
+  } else if (IsType<framework::SelectedRows>()) {
+    return GetMutable<framework::SelectedRows>()
+        ->mutable_value()
+        ->InplaceVersionCounter();
+  } else if (IsType<framework::LoDTensorArray>()) {
+    auto tensor_array = GetMutable<framework::LoDTensorArray>();
+    return tensor_array->at(0).InplaceVersionCounter();
+  } else {
+    // NOTE(liym27): Other types of data are not supported, like Scope.
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Only supports Tensor, LoDTensor, SelectedRows and "
+        "LoDTensorArray to have TensorInplaceVersion."));
```

Comment:
Can we give the current type in error message?

Reply:
Yes, thanks, done